pkg/daemon/update.go

```diff
@@ -512,58 +515,73 @@ func (dn *Daemon) writeUnits(units []ignv2_2types.Unit) error {
 	return nil
 }
 
-// writeFiles writes the given files to disk.
-// it doesn't fetch remote files and expects a flattened config file.
-func (dn *Daemon) writeFiles(files []ignv2_2types.File) error {
-	for _, f := range files {
-		glog.Infof("Writing file %q", f.Path)
-		// create any required directories for the file
-		if err := dn.fileSystemClient.MkdirAll(filepath.Dir(f.Path), DefaultDirectoryPermissions); err != nil {
-			return fmt.Errorf("Failed to create directory %q: %v", filepath.Dir(f.Path), err)
-		}
+func (dn *Daemon) atomicallyWriteFile(file ignv2_2types.File) error {
+	var succeded bool
 
-		// create the file
-		file, err := dn.fileSystemClient.Create(f.Path)
-		if err != nil {
-			return fmt.Errorf("Failed to create file %q: %v", f.Path, err)
+	// create any required directories for the file
+	if err := os.MkdirAll(filepath.Dir(file.Path), DefaultDirectoryPermissions); err != nil {
+		return fmt.Errorf("failed to create directory %q: %v", filepath.Dir(file.Path), err)
+	}
+
+	tempFile, err := ioutil.TempFile(filepath.Dir(file.Path), "file-write-tmp")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary file for %q", file.Path)
+	}
+	defer func() {
+		tempFile.Close()
+		if !succeded {
+			os.Remove(tempFile.Name())
 		}
-		defer file.Close()
+	}()
+
+	// write the file to disk, using the inlined file contents
+	contents, err := dataurl.DecodeString(file.Contents.Source)
+	if err != nil {
+		return err
+	}
+	if _, err = tempFile.WriteString(string(contents.Data)); err != nil {
+		return fmt.Errorf("failed to write inline contents to file %q: %v", tempFile.Name(), err)
+	}
 
-		// write the file to disk, using the inlined file contents
-		contents, err := dataurl.DecodeString(f.Contents.Source)
+	// chmod and chown
+	mode := DefaultFilePermissions
+	if file.Mode != nil {
+		mode = os.FileMode(*file.Mode)
+	}
+	if err := tempFile.Chmod(mode); err != nil {
+		return fmt.Errorf("failed to set file mode for file %q: %v", tempFile.Name(), err)
+	}
+
+	// set chown if file information is provided
+	if file.User != nil || file.Group != nil {
+		uid, gid, err := getFileOwnership(file)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to retrieve file ownership for file %q: %v", file.Path, err)
 		}
-		if _, err = file.WriteString(string(contents.Data)); err != nil {
```

Comment:
Discussed OOB but leaving a note here so the context isn't lost: this will misbehave as soon as the content is a pure binary blob that can't be represented as UTF-8.

Reply:
Along with that, we'll have to move to buffered writes as well, I'm spinning up another PR from this one (even if I do think we need this)

Reply:
https://github.com/openshift/machine-config-operator/pull/411